RestaurantsNearMe/Presentation/SplashScreen/SplashScreenViewModel.swift

```diff
@@ -5,38 +5,36 @@
 //  Created by Colin Evans on 2023-11-24.
 //
 
-// TODO: Figure out where to refresh account status then from there we can call fetch
-
-
 import Foundation
 import Combine
 import CloudKit
+import Networking
 
-class SplashScreenViewModel: ObservableObject {
-  private var isFetching = PassthroughSubject<Bool, Never>()
+class SplashScreenViewModel<Y: YelpAPIRequesting>: ObservableObject {
```

Comment:
Are you referring to `some MyProtocol` vs `<T: MyProtocol>`?

Reply:
yes, using `some` requires that the type is set in the declaration `var type: some MyProtocol = ConreteImplementation()`

When reading into these more, I believe the convention was to avoid existentials unless needed for their flexibility. In this case, since i'm injecting the dependency i'd either have to use `any` or generics. Is `any` the best practice over generics?

Reply:
`some` is just syntactic sugar, these two are equivalent:
```
func doSomething<T: MyProtocol>(with other: T)

// VS //

func doSomething(with other: some MyProtocol)
```

You only require these when `MyProtocol` has an `associatedtype`, otherwise you can just use `MyProtocol` directly.

```
class YelpAPIRequest: YelpAPIRequesting {
  private let requestBuilder: URLRequestBuilding
  private let cloudKitServiceProvider: CloudKitServiceProviding
  private var activeLocation: CLLocation?
  private var cancellables = Set<AnyCancellable>()
  
  init(
    requestBuilder: URLRequestBuilding,
    locationProvider: LocationProviding,
    cloudKitServiceProvider: CloudKitServiceProviding
  ) {
    self.requestBuilder = requestBuilder
    self.cloudKitServiceProvider = cloudKitServiceProvider
    locationProvider
      .currentLocation
      .sink { self.activeLocation = $0 }
      .store(in: &cancellables)
  }
  
  func updateAPIKey() {
    guard let privateKey = cloudKitServiceProvider.apiKey.value else {
      return
    }
    let fullAPIKey = "Bearer " + privateKey
    requestBuilder.withHeaderOptions(headerType: .auth, value: fullAPIKey)
  }
  
  func getUpdatedRestaurants() async {
    guard let location = activeLocation else {
      return
    }
    requestBuilder.addQueryItems(
      for:
        (name: "latitude", value: String(location.coordinate.latitude)),
        (name: "longitude", value: String(location.coordinate.longitude))
    )
    let apiRequest = requestBuilder.retrieveRequest()
    do {
     let data = try await URLSession.shared.data(for: apiRequest)
     print(data) // still need to parse data
    } catch {
      print(error)
    }
  }
}

extension YelpAPIRequest {
  static func preview() -> YelpRequest {
    YelpAPIRequest(
      requestBuilder: URLRequestBuilder(URL(string: "")!),
      locationProvider: LocationProvidingMock(),
      cloudKitServiceProvider: CloudKitServiceProvidingMock()
    )
  }
}
```

Reply:
so using `MyProtocol` directly doesn't cause the underlying variable to be wrapped in an existential container? it's only done so when the protocol has an associated type?

I guess I got confused from this exert from `Advanced-Swift`

```
The fact that Swift has been using the same syntax for the protocol (constraint) and the
existential type has contributed to the general confusion among Swift programmers
about the distinction between the two, especially because existentials have some
unintuitive limitations, as we’ll see in the following sections. This violates a
fundamental rule of good API and language design — namely that the same or similar
concepts should be spelled the same, and different things should be spelled differently.
It’s also unfortunate that the lightweight syntax might wrongly attract developers to a
feature (existentials) whose use is generally discouraged in favor of its syntax-heavier
alternatives (generics and opaque types) unless you specifically need the flexibility.
```

Reply:
I don't know the underpinnings well enough to comment about an existential container, but the way I understand it, if you have a protocol that is "augmented" by another one (e.g. `associatedtype`) then you need to use generics. (`<T>`) or existentials (`some`) to be more specific about which "version" of your base protocol you're using so the compiler can intuit the possible paths. If you don't care, or don't need to know, you can hint to the compiler that `any` version of the protocol will do.

Reply:
ahhh I see. So would you say that the best practices described in  ***Understanding the problems that any and some solve** https://www.donnywals.com/whats-the-difference-between-any-and-some-in-swift-5-7/ as overkill?

Reply:
That's an interesting article! I think the amount of protocols you're adding _might_ be overkill, if anything. The article does break down that `some` is better than `any` as the compiler can optimize things better, which is great, but there are times when you still _should_ use `any`.

The way I like to think about it is, the more protocols I add, the more mocks I likely need to create for testing. At some point, you're testing the mocks, rather than your business logic. I try to extract dependencies like `URLSession` and other things that would reduce testability.

It's probably worth asking yourself - what am I testing if I introduce this protocol?